CriticalMass/MapViewController.swift

```diff
@@ -221,6 +229,24 @@ extension MapViewController: MKMapViewDelegate {
             return nil
         }
 
+        guard annotation is CriticalMassAnnotation == false else {
+            if #available(iOS 11.0, *) {
+                if let cmAnnotationView = mapView.dequeueReusableAnnotationView(withIdentifier: CMMarkerAnnotationView.reuseIdentifier) {
+                    cmAnnotationView.annotation = annotation
+                    return cmAnnotationView
+                } else {
+                    return CMMarkerAnnotationView(annotation: annotation, reuseIdentifier: CMMarkerAnnotationView.reuseIdentifier)
+                }
+            } else {
+                if let cmAnnotationView = mapView.dequeueReusableAnnotationView(withIdentifier: CMAnnotationView.reuseIdentifier) {
+                    cmAnnotationView.annotation = annotation
+                    return cmAnnotationView
+                } else {
+                    return CMAnnotationView(annotation: annotation, reuseIdentifier: CMAnnotationView.reuseIdentifier)
+                }
+            }
+        }
+
```

Comment:
We should reuse the AnnotationController & AnnotationView infrastructure for that

Reply:
I will take a look at it

Reply:
Might be that my Generics knowledge does not reach far enough but I think Swifts Type inference is not able to add elements to `[AnnotationController]` if the other controller use different types then the `BikeAnnotation` and `BikeAnnotationView`

Reply:
I updated the `AnnotationController` since it did not needed the Generic constraints IMO

Reply:
Oh right :/ 

The only solution that came to my mind is adding a Protocol around it which looks like that

```swift
protocol AnnotationControllable {
    var annotationType: IdentifiableAnnnotation.Type { get }
    var annotationViewType: MKAnnotationView.Type { get }
}

class AnnotationController<T: IdentifiableAnnnotation, K: MKAnnotationView> {
    private let _annotationType = T.self
    private let _annotationViewType = K.self

    [...]
}

extension AnnotationController: AnnotationControllable {
    var annotationType: IdentifiableAnnnotation.Type {
        _annotationType
    }

    var annotationViewType: MKAnnotationView.Type {
        _annotationViewType
    }
}
```

And then use AnnotationControllable as a type. The generic constraint is needed to make sure at compile time that MKMapView+Register never fails. We can't safely make that assumption if the annotationType and annotationViewType is mutable

Reply:
But since it is a MKAnnotation and MKAnnotationView it will always register them and then we have to ask if the annotation is of type `Something` anyways, right?

Reply:
Yes it will also work, it will just break if you accidentally use the wrong initialiser for a given AttentionController. Another benefit of the Protocol is that all the code is in a single place and you don't need to override every initialiser

Reply:
I agree with the benefits of a protocol

Reply:
But there is only one init for the controller and the types are constraint to the specific controller. So I do not see where it should break. But I can rework to the protocol solution